### PR TITLE
Bugfix Catalyst: add GlobalPhase in TOML files

### PR DESCRIPTION
**Pull request by vincentmr**
…os and L-GPU.

### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Catalyst tests [fail](https://github.com/PennyLaneAI/catalyst/actions/runs/7929212681/job/21648998841#step:21:77333) with 
```
frontend/test/pytest/test_transform.py::test_batch_params[lightning.qubit] - catalyst.utils.exceptions.CompileError: Gates in qml.device.operations and specification file do not match
```
following the introduction of `GlobalPhase` support (#579).

**Description of the Change:**
Add `GlobalPhase` in TOML files.
Add tests that validate the adjoint method for all supported gates.
Add `GlobalPhase` generators in L-GPU and L-Kokkos.
 

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
Bug fix update Catalyst TOML file [sc-56992]

### CODE REVIEW

File: pennylane_lightning/lightning_qubit/lightning_qubit.toml [hunk 1]
```diff
@@ -116,7 +117,7 @@ runtime_code_generation = false
 # If the device supports adjoint
 quantum_adjoint = true
 # If the device supports quantum control instructions natively
-quantum_control = false
+quantum_control = true
```
Comment (erick-xanadu) [inline pennylane_lightning/lightning_qubit/lightning_qubit.toml:120]: @dime10 I know you suggested this. In the frontend, we do the following:

```python
        if QJITDevice._check_quantum_control(config):  # pragma: nocover
            QJITDevice.operations += ["QCtrl"]
```

I unconditionally added QCtrl, like so:

```python
        if QJITDevice._check_quantum_control(config):  # pragma: nocover
            QJITDevice.operations += ["QCtrl"]
        QJITDevice.operations += ["QCtrl"]
```

and I get this error:

```python
    def trace_quantum(self, ctx, device, trace, qrp) -> QRegPromise:
        raise NotImplementedError("QCtrl does not support JAX quantum tracing")  # pragma: no cover
```

Would you reconsider keeping false for a while longer? Or what do you suggest?

Comment (dime10) [reply pennylane_lightning/lightning_qubit/lightning_qubit.toml:120]: This is a deficiency on Catalyst's part I think, for the moment we should always decompose regardless of what the device says, since we have no representation for Control in MLIR.

The update should still take place here since the device now has the support. We'll catch up once Sergei finishes his implementation.

Comment (vincentmr) [reply pennylane_lightning/lightning_qubit/lightning_qubit.toml:120]: Ok, so we keep it `true` here, but Catalyst will disable it forcefully until it supports it. Alright?

Comment (erick-xanadu) [reply pennylane_lightning/lightning_qubit/lightning_qubit.toml:120]: I'm good with that if @dime10 is also good with that.

**Review by erick-xanadu — APPROVED** [review]

File: tests/test_adjoint_jacobian.py [hunk 2]
```diff
@@ -694,6 +694,64 @@ def circuit(x, y, z):
...
+            qml.IsingXY,
+            qml.IsingYY,
+            qml.IsingZZ,
+            qml.SingleExcitation,
+            qml.SingleExcitationMinus,
+            qml.SingleExcitationPlus,
+            qml.DoubleExcitation,
+            qml.DoubleExcitationMinus,
+            qml.DoubleExcitationPlus,
+            qml.GlobalPhase,
```
Comment (maliasadi) [inline tests/test_adjoint_jacobian.py:714]: is that the list of all supported operations? :thinking:

File: tests/test_adjoint_jacobian.py [hunk 3]
```diff
@@ -694,6 +694,64 @@ def circuit(x, y, z):
...
+        ],
+    )
+    @pytest.mark.parametrize("n_qubits", range(2, 6))
+    @pytest.mark.parametrize("par", [-np.pi / 7, np.pi / 5, 2 * np.pi / 3])
+    def test_gate_jacobian(self, par, n_qubits, operation, tol):
+        """Test that the jacobian of the controlled gate matches the parameter-shift formula."""
```
Comment (maliasadi) [inline tests/test_adjoint_jacobian.py:720]: ?
```suggestion
        """Test that the jacobian of the controlled gate matches the finite-diff formula."""
```

**Review by maliasadi — APPROVED** [review]
:raised_hands:

File: tests/test_adjoint_jacobian.py (hunk 2 above)
Comment (vincentmr) [reply tests/test_adjoint_jacobian.py:714]: Good spot, some implemented generators are missing here. Let me add them.

File: Makefile [hunk 4]
```diff
@@ -98,10 +98,10 @@ test-cpp:
 	cmake -BBuildTests -G Ninja -DCMAKE_BUILD_TYPE=Debug -DBUILD_TESTS=ON -DENABLE_WARNINGS=ON -DPL_BACKEND=$(PL_BACKEND) $(OPTIONS)
 ifdef target
 	cmake --build ./BuildTests $(VERBOSE) --target $(target)
-	OMP_PROC_BIND=false ./BuildTests/$(target)
+	./BuildTests/$(target)
 else
 	cmake --build ./BuildTests $(VERBOSE)
-	OMP_PROC_BIND=false cmake --build ./BuildTests $(VERBOSE) --target test
+	cmake --build ./BuildTests $(VERBOSE) --target test
```
Comment (mlxd) [inline Makefile:104]: Why is this removed? Don't we want to silence the Kokkos warning from this? Is the env var set elsewhere?

File: pennylane_lightning/core/src/simulators/lightning_kokkos/gates/tests/Test_StateVectorKokkos_Generator.cpp [hunk 5]
```diff
@@ -31,14 +35,15 @@
...
+using namespace Pennylane::Gates;
 using namespace Pennylane::LightningKokkos;
 using namespace Pennylane::Util;
 using std::size_t;
 } // namespace
 /// @endcond
 
-TEMPLATE_TEST_CASE("StateVectorKokkos::applyGenerator",
-                   "[StateVectorKokkosManaged_Generator]", float, double) {
+TEMPLATE_TEST_CASE("StateVectorKokkos::applyGenerator - errors",
```
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_kokkos/gates/tests/Test_StateVectorKokkos_Generator.cpp:45]: Why the "errors" label here?

File: pennylane_lightning/core/src/simulators/lightning_kokkos/gates/tests/Test_StateVectorKokkos_Generator.cpp [hunk 6]
```diff
@@ -48,1186 +53,79 @@ TEMPLATE_TEST_CASE("StateVectorKokkos::applyGenerator",
...
-                          .margin(EP));
-                CHECK(scale * real(result_gntr_sv[j]) ==
-                      Approx(0.5 *
-                             (imag(result_gate_svp[j]) -
-                              imag(result_gate_svm[j])) /
-                             ep)
-                          .margin(EP));
-            }
-        }
+TEMPLATE_TEST_CASE("StateVectorKokkos::applyGenerator",
```
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_kokkos/gates/tests/Test_StateVectorKokkos_Generator.cpp:56]: Good call on the generalisation here.

File: pennylane_lightning/lightning_qubit/lightning_qubit.toml (hunk 1 above)
Comment (mlxd) [inline pennylane_lightning/lightning_qubit/lightning_qubit.toml:120]: I'm curious how this makes sense to be true?

File: tests/test_adjoint_jacobian.py [hunk 7]
```diff
@@ -722,7 +785,7 @@ def circuit(x, y, z):
     @pytest.mark.parametrize("n_qubits", range(2, 6))
     @pytest.mark.parametrize("par", [-np.pi / 7, np.pi / 5, 2 * np.pi / 3])
     def test_controlled_jacobian(self, par, n_qubits, control_value, operation, tol):
-        """Test that the jacobian of the controlled gate matches the parameter-shift formula."""
+        """Test that the jacobian of the controlled gate matches the finite-diff formula."""
```
Comment (mlxd) [inline tests/test_adjoint_jacobian.py:788]: Should we be using FD here or PS? Given the noise introduced by FD, I'd say PS may be a better long-term option, but will leave it to be your call (e.g. changing L817 or leaving as is)

File: pennylane_lightning/lightning_qubit/lightning_qubit.toml (hunk 1 above)
Comment (erick-xanadu) [reply pennylane_lightning/lightning_qubit/lightning_qubit.toml:120]: There was a small discussion here: https://github.com/PennyLaneAI/pennylane-lightning/pull/615#discussion_r1492826930

The meaning `quantum_control` was roughly determined to be "supported gates can be receive an optional vector of qubits which correspond to control qubits". There was also a discussion on slack saying that the above meaning was now the case? Happy to learn if that is not the case.

File: Makefile (hunk 4 above)
Comment (vincentmr) [reply Makefile:104]: I found this to slow down the C++ tests considerably on many-core CPUs, so I thought I would just let the developer define the variable as he sees fit. The Kokkos warning does not annoy me that much, but I don't mind leaving it and unsetting myself when suitable.

File: pennylane_lightning/core/src/simulators/lightning_kokkos/gates/tests/Test_StateVectorKokkos_Generator.cpp (hunk 5 above)
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/gates/tests/Test_StateVectorKokkos_Generator.cpp:45]: I had to rename the test to avoid a name conflict. Because the test is only checking for a `LightningException` and I wanted to call the generic test `StateVectorKokkos::applyGenerator`, which was previously called `StateVectorKokkos::applyGeneratorXXX` with `XXX` the gate name, I renamed it with `errors`.

File: pennylane_lightning/lightning_qubit/lightning_qubit.toml (hunk 1 above)
Comment (vincentmr) [reply pennylane_lightning/lightning_qubit/lightning_qubit.toml:120]: I think the meaning is arbitrary controls (wires and values) are supported for all gates, which Lightning does with this PR.

Comment (AmintorDusko) [reply pennylane_lightning/lightning_qubit/lightning_qubit.toml:120]: It would be great to have a more meaningful name for this variable, and also some update to the docstring above.

File: tests/test_adjoint_jacobian.py (hunk 7 above)
Comment (vincentmr) [reply tests/test_adjoint_jacobian.py]: PS rules were missing for certain n-controlled gates, but they all seem to work fine now. I'll thus switch to PS. I'm keeping FD for `test_gate_jacobian` because PS fails for GlobalPhase.

File: Makefile (hunk 4 above)
Comment (mlxd) [reply Makefile:104]: I'm fine leaving it out then. :+1:

File: pennylane_lightning/lightning_qubit/lightning_qubit.toml (hunk 1 above)
Comment (erick-xanadu) [reply pennylane_lightning/lightning_qubit/lightning_qubit.toml:120]: I think changing the name is something we can do, but that is out of scope for this PR. That would affect other toml files stored in Catalyst. Thanks for the suggestion at @AmintorDusko! We will be having some changes in toml files in the coming weeks, specifically regarding `quantum_control` so we will keep it in mind.

Comment (mlxd) [reply pennylane_lightning/lightning_qubit/lightning_qubit.toml:120]: Thanks all. In the interest of keeping things maintainable, let's avoiding keeping docs on Slack and elsewhere and make sure they are documented herein where possible. For me, I'd have no idea how this is consumed based on the name alone, and the associated value. I'll mark as resolved, but we can plan to update docs in future.

File: tests/test_adjoint_jacobian.py (hunk 7 above)
Comment (mlxd) [reply tests/test_adjoint_jacobian.py]: Wait, why does it fail for PS and not FD? Is that an issue to resolve?

Comment (vincentmr) [reply tests/test_adjoint_jacobian.py]: According to the [doc](https://docs.pennylane.ai/en/stable/code/api/pennylane.GlobalPhase.html), there is no gradient recipe, so it means PS will fail no?

**Review by mlxd — APPROVED** [review]
Thanks @vincentmr 

As a follow up, we can track whether the PS recipe needs specific handling for the GlobalPhase input parameter (which shouldn't be trainable).
